pkg/kubelet/kubelet.go

```diff
@@ -1671,6 +1672,32 @@ func (kl *Kubelet) canAdmitPod(pods []*v1.Pod, pod *v1.Pod) (bool, string, strin
 	return true, "", ""
 }
 
+// HandleAdmissionFailure allows the kubelet to gracefully handle admission rejection, and, in some cases,
+// to allow admission of the pod despite its previous failure.
+func (kl *Kubelet) HandleAdmissionFailure(pod *v1.Pod, failureReasons []algorithm.PredicateFailureReason) (bool, []algorithm.PredicateFailureReason, error) {
+	if kubetypes.IsCriticalPod(pod) {
+		// InsufficientResourceError is not a reason to reject a critical pod.
+		// Instead of rejecting, we free up resources to admit it, if no other reasons for rejection exist.
+		nonResourceReasons := []algorithm.PredicateFailureReason{}
+		resourceReasons := []predicates.InsufficientResourceError{}
+		for _, reason := range failureReasons {
+
+			if r, ok := reason.(*predicates.InsufficientResourceError); ok {
+				resourceReasons = append(resourceReasons, *r)
+			} else {
+				nonResourceReasons = append(nonResourceReasons, reason)
+			}
+		}
+		if len(nonResourceReasons) == 0 {
+			err := kl.evictionManager.PreemptPods(resourceReasons)
+			// if no error is returned, preemption succeeded and the pod is safe to admit.
+			return err == nil, []algorithm.PredicateFailureReason{}, err
+		}
+		return false, nonResourceReasons, nil
```

Comment:
That is intentional.  It would be confusing if a critical pod failed listing an insufficient resource as the reason, since the goal here is to specifically handle insufficient resources for critical pods.